scheds/rust/scx_p2dq/src/bpf/main.bpf.c

```diff
@@ -0,0 +1,1006 @@
+		p->scx.dsq_vtime = vtime_now + vtime_delta;
+	}
+
+	u64 vtime = p->scx.dsq_vtime;
+
+	/*
+	 * Limit the amount of budget that an idling task can accumulate
+	 * to one slice for the dsq.
+	 */
+	if (time_before(vtime, vtime_now - slice_ns))
+		vtime = vtime_now - slice_ns;
+
+	p->scx.dsq_vtime = vtime;
+
+	/*
+	 * Push per-cpu kthreads at the head of local dsq's and preempt the
+	 * corresponding CPU. This ensures that e.g. ksoftirqd isn't blocked
+	 * behind other threads which is necessary for forward progress
+	 * guarantee as we depend on the BPF timer which may run from ksoftirqd.
+	 */
+	if ((p->flags & PF_KTHREAD) && p->nr_cpus_allowed < nr_cpus && kthreads_local) {
+		scx_bpf_dsq_insert(p, SCX_DSQ_LOCAL, slice_ns,
+				   enq_flags | SCX_ENQ_PREEMPT);
+		return;
+	}
+
+	/* 
+	 * Affinitized tasks just get dispatched directly, need to handle this better 
+	 */
+	if ((!taskc->all_cpus)) {
+		scx_bpf_dsq_insert(p, SCX_DSQ_LOCAL, slice_ns, enq_flags);
+	}
+
+	dsq_id = cpu_dsq_id(taskc->dsq_index, cpuc);
+	scx_bpf_dsq_insert_vtime(p, dsq_id, slice_ns, vtime, enq_flags);
+
+	if (dsq_id >= 0 && dsq_id < nr_dsqs_per_llc && vtime > llcx->dsq_max_vtime[dsq_id]) {
+		llcx->dsq_max_vtime[dsq_id] = vtime;
+		trace("LLC[%d] DSQ[%d] max_vtime %llu", llcx->id, dsq_id, vtime);
+	}
+}
+
+
+void BPF_STRUCT_OPS(p2dq_runnable, struct task_struct *p, u64 enq_flags)
+{
+	struct task_ctx *wakee_ctx;
+
+	if (!(wakee_ctx = lookup_task_ctx(p)))
+		return;
+
+	wakee_ctx->is_kworker = p->flags & PF_WQ_WORKER;
+	wakee_ctx->sum_runtime = 0;
+}
+
+
+void BPF_STRUCT_OPS(p2dq_running, struct task_struct *p)
+{
+	struct task_ctx *taskc;
+	struct cpu_ctx *cpuc;
+	struct llc_ctx *llcx;
+	s32 task_cpu = scx_bpf_task_cpu(p);
+
+	if (!(taskc = lookup_task_ctx(p)))
+		return;
+
+	if (!(cpuc = lookup_cpu_ctx(task_cpu)) || !(llcx = lookup_llc_ctx(cpuc->llc_id)))
+		return;
+
+	taskc->last_run_at = scx_bpf_now();
+	taskc->llc_id = cpuc->llc_id;
+	llcx->vtime = p->scx.dsq_vtime;
+	cpuc->dsq_index = taskc->dsq_index;
+	cpuc->ran_for = 0;
+	taskc->cpu = task_cpu;
+
+	// In perf mode give interactive tasks a perf boost.
+	if (sched_mode == MODE_PERFORMANCE || is_interactive(taskc)) {
+		cpuc->perf = 1024;
+		scx_bpf_cpuperf_set(task_cpu, cpuc->perf);
+	}
+}
+
+
+void BPF_STRUCT_OPS(p2dq_stopping, struct task_struct *p, bool runnable)
+{
+	struct task_ctx *taskc;
+	struct cpu_ctx *cpuc;
+	struct llc_ctx *llcx;
+	u64 used, last_dsq_slice_ns;
+
+	if (!(taskc = lookup_task_ctx(p)))
+		return;
+
+	if (!(cpuc = lookup_cpu_ctx(-1)) || !(llcx = lookup_llc_ctx(cpuc->llc_id)))
+		return;
+
+	u64 now = scx_bpf_now();
+	last_dsq_slice_ns = dsq_time_slice(taskc->dsq_index);
+	used = now - taskc->last_run_at;
+	taskc->last_run_at = now;
+	taskc->last_dsq_id = taskc->dsq_id;
+	cpuc->dsq_index = taskc->dsq_index;
+
+	// On stopping determine if the task can move to a longer DSQ by
+	// comparing the used time to the scaled DSQ slice.
+	if (used >= ((9 * last_dsq_slice_ns) / 10)) {
```

Comment:
Is it worth making this `7/8`s or `15/16`s to make the division easier? The 9/10ths seems arbitrary but perhaps there's a good reason.